aten/src/ATen/native/sparse/SparseMatMul.cpp

```diff
@@ -0,0 +1,255 @@
+#include <ATen/ATen.h>
+#include <ATen/Config.h>
+#include <ATen/NamedTensorUtils.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/Parallel.h>
+#include <ATen/SparseTensorImpl.h>
+#include <ATen/SparseTensorUtils.h>
+
+namespace at { namespace native {
+
+using namespace at::sparse;
+
+/*
+    This is an implementation of the SMMP algorithm:
+     "Sparse Matrix Multiplication Package (SMMP)"
+
+      Randolph E. Bank and Craig C. Douglas
+      https://doi.org/10.1007/BF02070824
+*/
+namespace {
+void csr_to_coo(const int64_t n_row, const int64_t Ap[], int64_t Bi[]) {
+  /* 
+    Expands a compressed row pointer into a row indices array
+    Inputs:
+      `n_row` is the number of rows in `Ap`
+      `Ap` is the row pointer
+    
+    Output: 
+      `Bi` is the row indices 
+  */
+  for (int64_t i = 0; i < n_row; i++) {
+    for (int64_t jj = Ap[i]; jj < Ap[i + 1]; jj++) {
+      Bi[jj] = i;
+    }
+  }
+}
+
+int64_t _csr_matmult_maxnnz(
+    const int64_t n_row,
+    const int64_t n_col,
+    const int64_t Ap[],
+    const int64_t Aj[],
+    const int64_t Bp[],
+    const int64_t Bj[]) {
+  /*
+    Compute needed buffer size for matrix `C` in `C = A*B` operation.
+
+    The matrices should be in proper CSR structure, and their dimensions
+    should be compatible.
+  */
+  std::vector<int64_t> mask(n_col, -1);
+  int64_t nnz = 0;
+  for (int64_t i = 0; i < n_row; i++) {
+    int64_t row_nnz = 0;
+
+    for (int64_t jj = Ap[i]; jj < Ap[i + 1]; jj++) {
+      int64_t j = Aj[jj];
+      for (int64_t kk = Bp[j]; kk < Bp[j + 1]; kk++) {
+        int64_t k = Bj[kk];
+        if (mask[k] != i) {
+          mask[k] = i;
+          row_nnz++;
+        }
+      }
+    }
+    int64_t next_nnz = nnz + row_nnz;
+    nnz = next_nnz;
+  }
+  return nnz;
+}
+
+template<class scalar_t>
+void _csr_matmult(
+    const int64_t n_row,
+    const int64_t n_col,
+    const int64_t Ap[],
+    const int64_t Aj[],
+    const scalar_t Ax[],
+    const int64_t Bp[],
+    const int64_t Bj[],
+    const scalar_t Bx[],
+    int64_t Cp[],
+    int64_t Cj[],
+    scalar_t Cx[]) {
+  /* 
+    Compute CSR entries for matrix C = A*B.
+
+    The matrices `A` and 'B' should be in proper CSR structure, and their dimensions
+    should be compatible.
+
+    Inputs:
+      `n_row`         - number of row in A
+      `n_col`         - number of columns in B
+      `Ap[n_row+1]`   - row pointer
+      `Aj[nnz(A)]`    - column indices
+      `Ax[nnz(A)]     - nonzeros
+      `Bp[?]`         - row pointer
+      `Bj[nnz(B)]`    - column indices
+      `Bx[nnz(B)]`    - nonzeros
+    Outputs:
+      `Cp[n_row+1]` - row pointer
```

Comment:
would "row index" be a better name than "row pointer"?

Reply:
I using these names because `row pointer`, `column indices` and `values` are commonly used to describe the properties of a matrix in `CSR` format.